op-e2e/actions/ecotone_fork_test.go

```diff
@@ -0,0 +1,139 @@
+	require.NoError(t, err)
+	t.Log("seq num", l1Info.SequenceNumber)
+
+	// L1Block: 1 set-L1-info + 2 deploys + 2 upgradeTo + 1 enable ecotone on GPO + 1 4788 deploy
+	// See [derive.EcotoneNetworkUpgradeTransactions]
+	require.Equal(t, 7, len(transactions))
+
+	// All transactions are successful
+	for i := 1; i < 7; i++ {
+		txn := transactions[i]
+		receipt, err := engine.EthClient().TransactionReceipt(context.Background(), txn.Hash())
+		require.NoError(t, err)
+		require.Equal(t, types.ReceiptStatusSuccessful, receipt.Status)
+	}
+
+	expectedL1BlockAddress := crypto.CreateAddress(derive.L1BlockDeployerAddress, 0)
+	expectedGasPriceOracleAddress := crypto.CreateAddress(derive.GasPriceOracleDeployerAddress, 0)
+
+	// Gas Price Oracle Proxy is updated
+	updatedGasPriceOracleAddress, err := engine.EthClient().StorageAt(context.Background(), predeploys.GasPriceOracleAddr, genesis.ImplementationSlot, latestBlock.Number())
+	require.NoError(t, err)
+	assert.Equal(t, expectedGasPriceOracleAddress, common.BytesToAddress(updatedGasPriceOracleAddress))
+	assert.NotEqualf(t, initialGasPriceOracleAddress, updatedGasPriceOracleAddress, "Gas Price Oracle Proxy address should have changed")
+
+	// L1Block Proxy is updated
+	updatedL1BlockAddress, err := engine.EthClient().StorageAt(context.Background(), predeploys.L1BlockAddr, genesis.ImplementationSlot, latestBlock.Number())
+	require.NoError(t, err)
+	assert.Equal(t, expectedL1BlockAddress, common.BytesToAddress(updatedL1BlockAddress))
+	assert.NotEqualf(t, initialL1BlockAddress, updatedL1BlockAddress, "L1Block Proxy address should have changed")
+
+	_, err = gasPriceOracle.Scalar(nil)
+	require.ErrorContains(t, err, "scalar() is deprecated")
+
+	cost, err := gasPriceOracle.GetL1Fee(nil, []byte{0, 1, 2, 3, 4})
+	require.NoError(t, err)
+	require.Equal(t, cost.Uint64(), uint64(0), "expecting zero scalars within activation block")
+
+	// 4788 contract is deployed
+	expected4788Address := crypto.CreateAddress(derive.EIP4788From, 0)
+	code, err := engine.EthClient().CodeAt(context.Background(), expected4788Address, latestBlock.Number())
+	require.NoError(t, err)
+	require.Equal(t, code, predeploys.EIP4788ContractCode)
+
+	// Build empty L2 block, to pass ecotone activation
+	sequencer.ActL2StartBlock(t)
+	sequencer.ActL2EndBlock(t)
+
+	// assert again, now that we are past activation
+	_, err = gasPriceOracle.Scalar(nil)
+	require.ErrorContains(t, err, "scalar() is deprecated")
+
+	// test if the migrated scalar matches the deploy config
+	basefeeScalar, err := gasPriceOracle.BaseFeeScalar(nil)
+	require.NoError(t, err)
+	require.True(t, uint64(basefeeScalar) == dp.DeployConfig.GasPriceOracleScalar, "must match deploy config")
+
+	cost, err = gasPriceOracle.GetL1Fee(nil, []byte{0, 1, 2, 3, 4})
+	require.NoError(t, err)
+	require.Greater(t, cost.Uint64(), uint64(0), "expecting non-zero scalars after activation block")
```

Comment:
why not test for the exact expected value given whatever the scalar is used by these tests?

Reply:
Unfortunately it looks like the GPO getL1-fee was not behaving correctly post-regolith pre-ecotone. I'll add a comment to it

Reply:
Hmm, looked closer into it, and it may be a feature, since the 68 bytes count towards signature/RLP overhead. Need to verify how the GPO is used here. The same 68 bytes is added by the GPO into the ecotone cost.

Reply:
Right, the GPO cost function assumes 68 non-zero bytes for the signature, so it won't exactly match the exact value computed by the "actual" function if there happens to be 0 bytes in the sig.